File: app/src/main/java/com/shreyaspatil/livestream/example/java/AsyncProcessor.java
package com.shreyaspatil.livestream.example.java;

import dev.shreyaspatil.livestream.LiveStream;

public class AsyncProcessor {

    private LiveStream<String> liveStream = new LiveStream<>();

    public void execute() {
        delay();
        liveStream.set("response", "Hi! Response from main thread");
    }

    public void executeInBackground() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                delay();
                liveStream.post("response", "Hi! Response from other thread");
            }
        }).start();
    }

    public void delay() {
        try {
            Thread.sleep(1000);
        } catch (Exception ignored) {
        }
    }
}

File: app/src/main/java/com/shreyaspatil/livestream/example/AsyncProcessor.kt
package com.shreyaspatil.livestream.example

import com.shreyaspatil.livestream.LiveStream
import kotlinx.coroutines.*

class AsyncProcessor {
    private val liveStream = LiveStream<String>()

    fun execute() {
        CoroutineScope(Dispatchers.Main).launch {
            val asyncTask = async{ asyncTask() }
            val response = asyncTask.await()

            // Emit network response.
            liveStream.emit("response", response)
        }
    }

    private suspend fun asyncTask(): String = coroutineScope {
        // Consider this block as a network call
        // Wait for some seconds
        delay(2000)

        // After some interval, store network response in result.
        val result = "Hi there! This is response from async function."
        result
    }
}
